teamcity: failed tests on master: testrace/TestLogic, testrace/TestLogic/default, testrace/TestLogic/default/views, testrace/TestLogic/parallel-stmts, testrace/TestLogic/parallel-stmts/views, testrace/TestLogic/distsql, testrace/TestLogic/distsql/views, test/TestLogic, test/TestLogic/default, test/TestLogic/default/views, test/TestLogic/parallel-stmts, test/TestLogic/parallel-stmts/views, test/TestLogic/distsql, test/TestLogic/distsql/views

The following tests appear to have failed:

[#245845](https://teamcity.cockroachdb.com/viewLog.html?buildId=245845):

```
--- FAIL: testrace/TestLogic (321.410s)
<autogenerated>:10: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic074306131
<autogenerated>:9: use -show-logs to present logs inline
--- FAIL: testrace/TestLogic/default (0.720s)
null
--- FAIL: testrace/TestLogic/default/views (17.620s)
logic_test.go:1718: 
	testdata/logic_test/views:360: expected "value type int\\[\\] cannot be used for table columns", but found <nil>
logic_test.go:1728: 
	testdata/logic_test/views:360: error in statement, skipping to next file
--- FAIL: testrace/TestLogic/parallel-stmts (0.800s)
null
--- FAIL: testrace/TestLogic/parallel-stmts/views (11.650s)
logic_test.go:1718: 
	testdata/logic_test/views:360: expected "value type int\\[\\] cannot be used for table columns", but found <nil>
logic_test.go:1728: 
	testdata/logic_test/views:360: error in statement, skipping to next file
--- FAIL: testrace/TestLogic/distsql (1.240s)
null
--- FAIL: testrace/TestLogic/distsql/views (28.220s)
logic_test.go:1718: 
	testdata/logic_test/views:360: expected "value type int\\[\\] cannot be used for table columns", but found <nil>
logic_test.go:1728: 
	testdata/logic_test/views:360: error in statement, skipping to next file
--- FAIL: test/TestLogic (27.880s)
test_log_scope.go:80: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic148844572
test_log_scope.go:63: use -show-logs to present logs inline
--- FAIL: test/TestLogic/default (0.120s)
null
--- FAIL: test/TestLogic/default/views (1.520s)
logic_test.go:1718: 
	testdata/logic_test/views:360: expected "value type int\\[\\] cannot be used for table columns", but found <nil>
logic_test.go:1728: 
	testdata/logic_test/views:360: error in statement, skipping to next file
--- FAIL: test/TestLogic/parallel-stmts (0.110s)
null
--- FAIL: test/TestLogic/parallel-stmts/views (1.370s)
logic_test.go:1718: 
	testdata/logic_test/views:360: expected "value type int\\[\\] cannot be used for table columns", but found <nil>
logic_test.go:1728: 
	testdata/logic_test/views:360: error in statement, skipping to next file
--- FAIL: test/TestLogic/distsql (0.160s)
null
--- FAIL: test/TestLogic/distsql/views (1.600s)
logic_test.go:1718: 
	testdata/logic_test/views:360: expected "value type int\\[\\] cannot be used for table columns", but found <nil>
logic_test.go:1728: 
	testdata/logic_test/views:360: error in statement, skipping to next file

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/sql/sqlbase/table.go
+++ b/pkg/sql/sqlbase/table.go
@@ -79,6 +79,16 @@ func MakeColumnDefDescs(
 		Nullable: d.Nullable.Nullability != parser.NotNull && !d.PrimaryKey,
 	}
 
+	// Prevent unsupported array types from falling through.
+	// TODO(jordan): This is ugly and a workaround for incompatibility between
+	// CastTargetToDatumType and structured.go's DatumTypeToColumnType. See #15813
+	switch t := d.Type.(type) {
+	case *parser.ArrayColType:
+		if _, ok := t.ParamType.(*parser.IntColType); !ok {
+			return nil, nil, errors.Errorf("arrays of type %s are unsupported", t.ParamType)
+		}
+	}
+
 	// Set Type.Kind and Type.Locale.
 	colDatumType := parser.CastTargetToDatumType(d.Type)
 	col.Type = DatumTypeToColumnType(colDatumType)

--- a/pkg/sql/testdata/logic_test/views
+++ b/pkg/sql/testdata/logic_test/views
@@ -357,5 +357,16 @@ CREATE TABLE t2 AS SELECT d, t FROM t AS OF SYSTEM TIME '2017-02-13 21:30:00'
 statement ok
 DROP TABLE t CASCADE
 
-statement error value type int\[\] cannot be used for table columns
-CREATE VIEW fail AS SELECT ARRAY[2]
+statement error cannot make array for column type BOOL
+CREATE VIEW fail AS SELECT ARRAY[true]
+
+statement error arrays of type STRING are unsupported
+CREATE VIEW fail AS SELECT ARRAY['foo']
+
+statement ok
+CREATE VIEW arr(a) AS SELECT ARRAY[3]
+
+query TI
+SELECT *, a[1] FROM arr
+----
+{3}  3

